diff -rup a/ipc.c b/ipc.c
--- a/ipc.c	2009-10-08 17:33:15.000000000 -0600
+++ b/ipc.c	2012-08-06 10:28:19.053314174 -0600
@@ -340,7 +340,11 @@ int sys_semtimedop(struct tcb *tcp)
 		if (indirect_ipccall(tcp)) {
 			tprint_sembuf(tcp, tcp->u_arg[3], tcp->u_arg[1]);
 			tprintf(", ");
-			printtv(tcp, tcp->u_arg[5]);
+#if defined(S390) || defined(S390X)
+                       printtv(tcp, tcp->u_arg[2]);
+#else
+                       printtv(tcp, tcp->u_arg[5]);
+#endif
 		} else {
 			tprint_sembuf(tcp, tcp->u_arg[1], tcp->u_arg[2]);
 			tprintf(", ");
