--- a/mem.c	2014-04-07 10:46:17.000000000 -0600
+++ b/mem.c	2014-04-07 10:51:47.410660692 -0600
@@ -401,6 +401,18 @@ static const struct xlat madvise_cmds[]
 #ifdef MADV_DONTNEED
 	{ MADV_DONTNEED,	"MADV_DONTNEED" },
 #endif
+#ifndef MADV_DODUMP
+#define MADV_DODUMP 17
+#endif
+#ifdef MADV_DODUMP
+	{ MADV_DODUMP,		"MADV_DODUMP" },
+#endif
+#ifndef MADV_DONTDUMP
+#define MADV_DONTDUMP 16
+#endif
+#ifdef MADV_DONTDUMP
+	{ MADV_DONTDUMP,	"MADV_DONTDUMP" },
+#endif
 	{ 0,			NULL },
 };
 
