Only in b: autom4te.cache
diff -rup a/config.h.in b/config.h.in
--- a/config.h.in	2016-05-31 07:35:56.000000000 -0400
+++ b/config.h.in	2017-05-12 17:32:49.134113672 -0400
@@ -1782,9 +1782,13 @@
 /* Define to 1 if the system has the type `struct blk_user_trace_setup'. */
 #undef HAVE_STRUCT_BLK_USER_TRACE_SETUP
 
-/* Define to 1 if `start struct btrfs_ioctl_search_args_v2.buf_size' is a
-   member of `struct btrfs_ioctl_defrag_range_args'. */
-#undef HAVE_STRUCT_BTRFS_IOCTL_DEFRAG_RANGE_ARGS_START_STRUCT_BTRFS_IOCTL_SEARCH_ARGS_V2_BUF_SIZE
+/* Define to 1 if `start' is a member of `struct
+   btrfs_ioctl_defrag_range_args'. */
+#undef HAVE_STRUCT_BTRFS_IOCTL_DEFRAG_RANGE_ARGS_START
+
+/* Define to 1 if `buf_size' is a member of `struct
+   btrfs_ioctl_search_args_v2'. */
+#undef HAVE_STRUCT_BTRFS_IOCTL_SEARCH_ARGS_V2_BUF_SIZE
 
 /* Define to 1 if `compat_flags' is a member of `struct
    btrfs_ioctl_feature_flags'. */
diff -rup a/configure b/configure
--- a/configure	2017-05-12 16:59:22.000000000 -0400
+++ b/configure	2017-05-12 17:15:19.843076820 -0400
@@ -8454,13 +8454,24 @@ _ACEOF
 
 
 fi
-ac_fn_c_check_member "$LINENO" "struct btrfs_ioctl_defrag_range_args" "start struct btrfs_ioctl_search_args_v2.buf_size" "ac_cv_member_struct_btrfs_ioctl_defrag_range_args_start_struct_btrfs_ioctl_search_args_v2_buf_size" " #include <stdio.h>
+ac_fn_c_check_member "$LINENO" "struct btrfs_ioctl_defrag_range_args" "start" "ac_cv_member_struct_btrfs_ioctl_defrag_range_args_start" " #include <stdio.h>
 #include <linux/btrfs.h>
 "
-if test "x$ac_cv_member_struct_btrfs_ioctl_defrag_range_args_start_struct_btrfs_ioctl_search_args_v2_buf_size" = xyes; then :
+if test "x$ac_cv_member_struct_btrfs_ioctl_defrag_range_args_start" = xyes; then :
 
 cat >>confdefs.h <<_ACEOF
-#define HAVE_STRUCT_BTRFS_IOCTL_DEFRAG_RANGE_ARGS_START_STRUCT_BTRFS_IOCTL_SEARCH_ARGS_V2_BUF_SIZE 1
+#define HAVE_STRUCT_BTRFS_IOCTL_DEFRAG_RANGE_ARGS_START 1
+_ACEOF
+
+
+fi
+ac_fn_c_check_member "$LINENO" "struct btrfs_ioctl_search_args_v2" "buf_size" "ac_cv_member_struct_btrfs_ioctl_search_args_v2_buf_size" " #include <stdio.h>
+#include <linux/btrfs.h>
+"
+if test "x$ac_cv_member_struct_btrfs_ioctl_search_args_v2_buf_size" = xyes; then :
+
+cat >>confdefs.h <<_ACEOF
+#define HAVE_STRUCT_BTRFS_IOCTL_SEARCH_ARGS_V2_BUF_SIZE 1
 _ACEOF
 
 
diff -rup a/configure.ac b/configure.ac
--- a/configure.ac	2017-05-12 16:59:22.000000000 -0400
+++ b/configure.ac	2017-05-12 17:15:13.077083844 -0400
@@ -464,7 +464,7 @@ AC_CHECK_HEADERS([linux/btrfs.h], [
 	AC_CHECK_MEMBERS(m4_normalize([
 		struct btrfs_ioctl_feature_flags.compat_flags,
 		struct btrfs_ioctl_fs_info_args.nodesize,
-		struct btrfs_ioctl_defrag_range_args.start
+		struct btrfs_ioctl_defrag_range_args.start,
 		struct btrfs_ioctl_search_args_v2.buf_size
 		]),,, [ #include <stdio.h>
 #include <linux/btrfs.h>])
