Index: strace-4.5.19/strace.1
===================================================================
--- strace-4.5.19.orig/strace.1
+++ strace-4.5.19/strace.1
@@ -43,7 +43,7 @@ strace \- trace system calls and signals
 .SH SYNOPSIS
 .B strace
 [
-.B \-dffhiqrtttTvxx
+.B \-dDffhiqrtttTvVxx
 ]
 [
 .BI \-a column
@@ -84,6 +84,9 @@ strace \- trace system calls and signals
 .B strace
 .B \-c
 [
+.B \-D
+]
+[
 .BI \-e expr
 ]
 \&...
@@ -243,6 +246,13 @@ program exit.  On Linux, this attempts t
 running in the kernel) independent of wall clock time.  If -c is used with
 -f or -F (below), only aggregate totals for all traced processes are kept.
 .TP
+.B \-D
+(Not available on SVR4 and FreeBSD.)
+Run tracer process as a detached grandchild, not as parent of the
+tracee.  This reduces the visible effect of
+.B strace
+by keeping the tracee a direct child of the calling process.
+.TP
 .B \-d
 Show some debugging output of
 .B strace
Index: strace-4.5.19/strace.c
===================================================================
--- strace-4.5.19.orig/strace.c
+++ strace-4.5.19/strace.c
@@ -170,10 +170,10 @@ FILE *ofp;
 int exitval;
 {
 	fprintf(ofp, "\
-usage: strace [-dffhiqrtttTvVxx] [-a column] [-e expr] ... [-o file]\n\
+usage: strace [-dDffhiqrtttTvVxx] [-a column] [-e expr] ... [-o file]\n\
               [-p pid] ... [-s strsize] [-u username] [-E var=val] ...\n\
               [command [arg ...]]\n\
-   or: strace -c -D [-e expr] ... [-O overhead] [-S sortby] [-E var=val] ...\n\
+   or: strace -c [-D] [-e expr] ... [-O overhead] [-S sortby] [-E var=val] ...\n\
               [command [arg ...]]\n\
 -c -- count time, calls, and errors for each syscall and report summary\n\
 -f -- follow forks, -ff -- with output into separate files\n\
@@ -736,7 +736,6 @@ main(int argc, char *argv[])
 			debug++;
 			break;
 #ifndef USE_PROCFS
-		/* Experimental, not documented in manpage yet. */
 		case 'D':
 			daemonized_tracer = 1;
 			break;
