Index: strace-4.5.19/process.c
===================================================================
--- strace-4.5.19.orig/process.c
+++ strace-4.5.19/process.c
@@ -906,6 +906,11 @@ struct tcb *tcp;
 	if (entering(tcp)) {
 		if (!followfork)
 			return 0;
+#ifdef LINUX
+		if (sysent[tcp->scno].sys_func == sys_clone &&
+		    (tcp->u_arg[ARG_FLAGS] & CLONE_UNTRACED))
+			return 0;
+#endif
 		fork_tcb(tcp);
 		if (setbpt(tcp) < 0)
 			return 0;
